### ISSUE TITLE
Fully synced node gets stuck after a few hours

### ISSUE THREAD

**Issue by torokati44** [iotaledger/iri#428]
The node is perfectly in sync for hours, then it just stops at a random milestone for no apparent reason, without being touched at all. The latestMilestone keeps advancing with COO, the latestSolidMilestone is stuck. It doesn't move at all for hours, but after an IRI restart, it will get back to full sync within a few minutes. Then another few hours pass, and it gets stuck again.

IRI version 1.4.1.4.
I have a couple manual neighbors, but also run Nelson (0.2.2).

**Comment by torokati44** [comment]
At the moment it is 308589 / 308546, if this helps.

**Comment by fijter** [comment]
Same here, with or without Nelson. Node has good neighbors but has been stuck at 243000 for hours now after a reboot.

**Comment by mstrewe** [comment]
Can approve this. But is not stucking.. its hanging behind. I follow my node stats and after ~ 20 minutes the number increase from 308638 to 308639... I will look tomorrow if its increasing

**Comment by mstrewe** [comment]
@fijter was your node already in sync?

**Comment by fijter** [comment]
My node was fine and always in sync up until the moment the COO got paused, after that I had issues. After a couple of restarts my node is in sync again but it didn't move by itself.

**Comment by stockermans** [comment]
I seem to have some issue as well. My node was in sync always with my neighbors, until the upgrade it seems. I had it staying updating the solidSubtangleMilestoneIndex like clockwork for a while yesterday (maybe a couple hours), and then it just stopped, and only the latestMilestoneIndex would continue, but my SolidSubtangleMilestone does not want to move.

I've tried with a fresh db from iota.partners and it will get in sync momentarily, then just will not continue.

I do not use nelson, only manual tethered neighbors which have been stable for a month+.

**Comment by mstrewe** [comment]
OK. Now I can confirm that subtangle stopped completely.

**Comment by scott79** [comment]
Adding some additional info found. Still need time to review further, however unsure if approach listed at end would make sense or cause other issues.

3 nodes

1 - consistently synced with latest milestone

2 - Not synced:

	1 - Public node and is connected to fully synced node
	1 - Private node connected to not synced public node

	Notes on non-synced nodes:

	- Upon clearing of db (on both nodes) and starting up process for non-synced public/private nodes. Public node requested transactions at a rate of about 1-2MB/sec from synced public node. The duration lasted
	for about 2 hours then dropped down to roughly 20kB/sec. 
	- Both non-synced nodes stayed in synced with respect to maintaining the same solid milestone index. 
	- Private node was restarted and solid milestone index jumped and is currently static at 308588. Its public peer is still updating solid milestone index at around the 269000 range. 
	Nodes are being left in this state until public node reaches 308588. 
	- Although private node milestone is static, disk usage tends to approach same disk usage as its peered node
		Public node: 	3819940	mainnetdb (Milestone advancing; currently at around 269900)
		Private node: 	3633920	mainnetdb (Milestone static @ 308588)

	After reviewing some of the code and adding line statements. I see something is causing the conditional statements in updateLatestSolidSubtangleMilestone method to return true and 
	setting latestSolidSubtangleMilestoneIndex.

	Output of logs:

	12/21 11:46:19.528 [Solid Milestone Tracker] INFO  com.iota.iri.Milestone - Checking milestone => 308589: LastSolidSubtangleMilestoneIndex: 308588
	12/21 11:46:19.776 [Solid Milestone Tracker] INFO  com.iota.iri.Milestone - Previous milestone => : 308587 solid: false

	However looking at all milestones prior, I see gaps in milestone solidity:

	12/21 11:57:18.980 [Solid Milestone Tracker] INFO  com.iota.iri.Milestone - Validating solidity of milestone: 258996 solid: true
	12/21 11:57:18.983 [Solid Milestone Tracker] INFO  com.iota.iri.Milestone - Validating solidity of milestone: 258997 solid: false

	-- Jumped to next section of non-solid milestones -- 
	12/21 11:57:20.747 [Solid Milestone Tracker] INFO  com.iota.iri.Milestone - Validating solidity of milestone: 259060 solid: false
	12/21 11:57:21.068 [Solid Milestone Tracker] INFO  com.iota.iri.Milestone - Validating solidity of milestone: 259061 solid: true


	The initial value of 'latestSolidSubtangleMilestoneIndex' is equal to the value ofand should be assigned to previousSolidSubtangleLatestMilestoneIndex. Given latestSolidSubtangleMilestoneIndex
	is less  than latestMilestoneIndex, it should execute the updateLatestSolidSubtangleMilestone() method. This method should stop at milestone 258997 
	(all milestones prior to this was true) but fails to.


	https://github.com/iotaledger/iri/blob/501e9493bc0cbd1ef1b3e0df6da39efb8ef4083d/src/main/java/com/iota/iri/LedgerValidator.java#L200

	However looking further, I see ledgerValidator::init actually updates milestone.latestSolidSubtangleMilestoneIndex = latestConsistentnMilestone.index(). After following the code 
	for ledgerValidator::buildSnapot method, it appears things are coming out of order. (Unsure this is expected)

	12/21 13:28:28.682 [Solid Milestone Tracker] INFO  com.iota.iri.LedgerValidator - Checking milestone snapshot: 243001
	12/21 13:28:28.730 [Solid Milestone Tracker] INFO  com.iota.iri.LedgerValidator - Milestone 243001 :snapshot consistent: true
	12/21 13:28:28.747 [Solid Milestone Tracker] INFO  com.iota.iri.LedgerValidator - Checking milestone snapshot: 243032
	12/21 13:28:28.774 [Solid Milestone Tracker] INFO  com.iota.iri.LedgerValidator - Milestone 243032 :snapshot consistent: true
	12/21 13:28:28.780 [Solid Milestone Tracker] INFO  com.iota.iri.LedgerValidator - Checking milestone snapshot: 243042
	12/21 13:28:28.815 [Solid Milestone Tracker] INFO  com.iota.iri.LedgerValidator - Milestone 243042 :snapshot consistent: true
	12/21 13:28:28.842 [Solid Milestone Tracker] INFO  com.iota.iri.LedgerValidator - Checking milestone snapshot: 243288
	12/21 13:28:28.874 [Solid Milestone Tracker] INFO  com.iota.iri.LedgerValidator - Milestone 243288 :snapshot consistent: true


	The buildsnapshot is ran initially and updates Milestone.latestSolidSubtangleMilestoneIndex with the most recent up to date consistent milestone snapshot index. Following the trail
	it appears the following method may be leading up to our inconsistency:

    		public MilestoneViewModel nextWithSnapshot(Tangle tangle) throws Exception {
        		MilestoneViewModel milestoneViewModel = next(tangle);
        		while(milestoneViewModel !=null && !StateDiffViewModel.exists(tangle, milestoneViewModel.getHash())) {
            			milestoneViewModel = milestoneViewModel.next(tangle);
        		}
        		return milestoneViewModel;
    		}


	12/21 13:52:18.054 [Solid Milestone Tracker] INFO  c.i.i.controllers.MilestoneViewModel - Current milestone => com.iota.iri.model.IntegerIndex@6d00d621 next => 243002
	12/21 13:52:18.054 [Solid Milestone Tracker] INFO  c.i.i.controllers.MilestoneViewModel - Current milestone => com.iota.iri.model.IntegerIndex@11e32dd0 next => 243003
	12/21 13:52:18.054 [Solid Milestone Tracker] INFO  c.i.i.controllers.MilestoneViewModel - Current milestone => com.iota.iri.model.IntegerIndex@7b14323f next => 243004
	12/21 13:52:18.054 [Solid Milestone Tracker] INFO  c.i.i.controllers.MilestoneViewModel - Current milestone => com.iota.iri.model.IntegerIndex@5cc9e0d1 next => 243005
	12/21 13:52:18.054 [Solid Milestone Tracker] INFO  c.i.i.controllers.MilestoneViewModel - Current milestone => com.iota.iri.model.IntegerIndex@379394b7 next => 243006
	12/21 13:52:18.054 [Solid Milestone Tracker] INFO  c.i.i.controllers.MilestoneViewModel - Current milestone => com.iota.iri.model.IntegerIndex@3f16b238 next => 243007
	12/21 13:52:18.054 [Solid Milestone Tracker] INFO  c.i.i.controllers.MilestoneViewModel - Current milestone => com.iota.iri.model.IntegerIndex@6549fa2f next => 243008
	12/21 13:52:18.054 [Solid Milestone Tracker] INFO  c.i.i.controllers.MilestoneViewModel - Current milestone => com.iota.iri.model.IntegerIndex@7615413f next => 243009
	12/21 13:52:18.055 [Solid Milestone Tracker] INFO  c.i.i.controllers.MilestoneViewModel - Current milestone => com.iota.iri.model.IntegerIndex@5d885566 next => 243010


	Verified the returned milestones are in sequential order which implys the StateDiffViewModel.exists call is the issue. Looking at this class, I see:

	 public static boolean exists(Tangle tangle, Hash hash) throws Exception {
        	return tangle.maybeHas(StateDiff.class, hash);
	}

	Given the need for consistency, would adding the transactionvalidator and checking for solidity work inside the nextWithSnapshot call?

**Comment by torokati44** [comment]
Stuck again. :)

**Comment by mstrewe** [comment]
now 24 hours full in sync...

**Comment by chrishamm** [comment]
Same problem here and I tried both OpenJDK 8 and Oracle JRE 8. My full node on CentOS 7 stays in sync for a few hours but then the last solid milestone is not updated any more, although I've got a few good neighbors. I can provide more details for debugging if requested.

**Comment by tiresias54** [comment]
same here, i have to restart IRI periodically to remain fully synchronized.

**Comment by mindlapse** [comment]
Same issue.
My node is connected to a second node that I own that is in sync, but this doesn't allow it to recover.
Running IRI 1.4.1.4 on both nodes.  5 neighbors total (one is mine), with 2GB memory and 2 cores.   It's running Oracle Java 1.8.0_151.

**Comment by torokati44** [comment]
I'm not paying for the VPS hosting my node anymore, until this is sorted out. Let me know when it happens.

**Comment by mindlapse** [comment]
I have two nodes, both running 1.4.1.4.  The one node is more powerful and is in sync.  The other node is connected to it but does not stay in sync.  They both have <= 5 neighbors.  I am downgrading the slower node to 1.4.1.2 for it to stay in sync.

Is there a way that the IRI full node can request the transactions it needs to complete the subtangle it is currently working on?

**Comment by toddmera** [comment]
Same here.  When I start IRI it sync completely.  However, after a few hours it gets behind and then never catches up.  When I restart IRI it catches up quickly and then after a few hours falls behind and does not catch up.  It actually never sycns from the point where it got behind.  Using IRI 1.4.1.4.  Plenty of CPU, RAM and Network.  This would also occur with IRI v 1.4.1.2.

**Comment by sreeraj-krishnan** [comment]
stuck again, have pulled changes 432 and 435
{
  "appName": "IRI",
  "appVersion": "1.4.1.4",
  "jreAvailableProcessors": 4,
  "jreFreeMemory": 1245916728,
  "jreVersion": "1.8.0_151",
  "jreMaxMemory": 9544663040,
  "jreTotalMemory": 2448949248,
  "latestMilestone": "YJ9WRGHKDOLUVKYOGEFXUMIIEXOQI9NNZRGKKUFURKYZIMPMQYWHBDEJVKKIYNVUIEKFOXSHRPUAZ9999",
  "latestMilestoneIndex": 311550,
  "latestSolidSubtangleMilestone": "EUVXVFCLBVJQSNSCWJEUQYWPJJSIPSZXGIEQFHDMQJNOQZKMVJVAODWVAEPPDSJLYKURM9RYHEP999999",
  "latestSolidSubtangleMilestoneIndex": 311540,
  "neighbors": 8,
  "packetsQueueSize": 0,
  "time": 1514212083245,
  "tips": 5000,
  "transactionsToRequest": 4,
  "duration": 0
}

**Comment by enewhuis** [comment]
Same here, stuck at 243006.  Not sure what I should do.  Slack users had mentioned one bad neighbor can cause this.  I'm using only Nelson so...I was hoping for hands-free.

**Comment by zenmetsu** [comment]
This isn't an issue with nelson.  

When you start the node, the starting milestone will be 243000 because that is where the last snapshot was taken.  The DB you downloaded, if you did, should contain the data since that milestone.  My understanding is that the DB tarballs are created hourly, so it should be almost up to date.

After starting the node, the code ingests all of that data (currently over 9GB) and does some work on it, possibly indexing, I haven't dug into it much.  You'll see possibly high CPU usage or high storage usage until this step is complete.  CPU, RAM, or storage can bottleneck your node during this point.  Once the DB processing is complete, if you are connected to functional neighbors, you should start to see milestone updates in the console output.  

Typically, a powerful node will take about 30 minutes to get to the point where it starts synchronizing.  Most reports from slack tend to indicate 50 minutes to just over an hour.  If you don't see milestone updates after 2 hours, I would take a good look at resource utilization on your node and see what is choking it.  Worst case scenario, blow out mainnetdb contents and download a fresh DB in the case that it is corrupted.

**Comment by vctender** [comment]
I have the same problem, node is never in sync mode anyway. I can see the Milestone changes in log, but I do not see a "Latest Solid Milestone Index" entry for hours. Even with re-installing using the pre-downloaded database.

I stuck know on solid milestone 312840. If I would restart the node, this value would increase too but stuck then again.

`12/27 10:10:01.462 [pool-2-thread-5] INFO  c.i.iri.network.TransactionRequester - Removed existing tx from request list: KBWWBSXPNOSEIOQ9TDF9QYLKJNQPFABJ9DKDZKHTYVWBXBXCNUPANYFJGWAMRTUYDEMBLBDKAJMQZ9999
12/27 10:10:01.485 [pool-2-thread-2] INFO  com.iota.iri.network.Node - toProcess = 0 , toBroadcast = 0 , toRequest = 1 , toReply = 0 / totalTransactions = 4284410
12/27 10:10:11.486 [pool-2-thread-2] INFO  com.iota.iri.network.Node - toProcess = 0 , toBroadcast = 0 , toRequest = 0 , toReply = 0 / totalTransactions = 4284422
12/27 10:10:21.487 [pool-2-thread-2] INFO  com.iota.iri.network.Node - toProcess = 0 , toBroadcast = 0 , toRequest = 0 , toReply = 0 / totalTransactions = 4284444
12/27 10:10:36.488 [pool-2-thread-2] INFO  com.iota.iri.network.Node - toProcess = 0 , toBroadcast = 0 , toRequest = 0 , toReply = 0 / totalTransactions = 4284469
12/27 10:10:46.510 [Latest Milestone Tracker] INFO  com.iota.iri.Milestone - Latest milestone has changed from #312877 to #312878
12/27 10:10:51.489 [pool-2-thread-2] INFO  com.iota.iri.network.Node - toProcess = 0 , toBroadcast = 0 , toRequest = 0 , toReply = 0 / totalTransactions = 4284497
`

**Comment by zenmetsu** [comment]
@vctender just posted an example of what is being discussed in this issue.  My prior comment was meant to point out that getting stuck @243000 is a separate issue entirely.

Would be nice to get some dev feedback as far as what needs to be collected... jstack,etc.  I have gobs of data collected during my attempts to troubleshoot this issue.

What I can say is that sometimes a client (wallet) will connect and do something like attach an address, and take down the solid subtangle processing.  Normally attaching an address takes a couple of seconds, but every now and then it will appear to hang for over minute.  When this happens, I notice that I don't get milestone updates on the console.  After the address is finally attached, I'll sometimes see 2 or 3 milestone updates in rapid succession.

Someone stated in another thread to not set heap, and to just let IRI take what it needs.  I tried that, and it went into swap (on a system with 24GB RAM).  At one point, full GC was triggered, it took 11 seconds, and after that the system stopped processing solid milestones entirely.  So it looks like if the system is taxed too hard, something breaks internally and the node will need to be restarted.

**Comment by vctender** [comment]
Nevertheless your problem may is related to pull request https://github.com/iotaledger/iri/pull/435

**Comment by sreeraj-krishnan** [comment]
I have #435 change and it does not fix it

**Comment by tawago** [comment]
Mine got fixed with #435.
No nelson, 5 neighbors were either fully synced or in the same situation.
1 core, 4GB ram.
LSMI always got stuck at some points a few hours after being successfully synced with LMI and #botbox. Not anymore with the pull request

**Comment by Berndinox** [comment]
I still got that issue and alot of people aswell, this is critical.
Any Solutions? 

.. been posted since a week now.

**Comment by OMSKremer** [comment]
Same problem here. No nelson, fullnode is installed via the howto from iota.partners on a new/clean ubuntu 16.04 LTS vm

**Comment by vctender** [comment]
Confirm: I compiled the current version including the pull request and the problem is not solved.

**Comment by sreeraj-krishnan** [comment]
Problem is not solved with #435.

Im running Nelson and iri does not sync. Max neighbours is 7.

**Comment by jdevelop** [comment]
Same here, latest build from `06650e4b7818f80344d42ae7dcc359fda039cbe7`.

At some point `"latestSolidSubtangleMilestoneIndex": 313782` is stuck for hours, upon node restart - it just jumps forward to match the `latestMilestoneIndex` once the latter is updated.

```
{
  "appName": "IRI",
  "appVersion": "1.4.1.4",
  "jreAvailableProcessors": 2,
  "jreFreeMemory": 779716984,
  "jreVersion": "1.8.0_152",
  "jreMaxMemory": 3711959040,
  "jreTotalMemory": 2008023040,
  "latestMilestone": "XYLSXZUAFEIUDVFTMPVHTHDRTCHNV9XUCYDNHBZJXIDNJATNESRCXLTHP9XOYVGIOTKF9MEYH9KP99999",
  "latestMilestoneIndex": 313799,
  "latestSolidSubtangleMilestone": "JP9XZPKSODKGEOLYJKZPNORQZDXMQGDFEDBGJQNXLWLUAVLMKXBLTV9EHIBSHXBAXWUMKTMDEPIKA9999",
  "latestSolidSubtangleMilestoneIndex": 313782,
  "neighbors": 8,
  "packetsQueueSize": 0,
  "time": 1514473298324,
  "tips": 38,
  "transactionsToRequest": 5,
  "duration": 1
}
```

**Comment by vctender** [comment]
I got now my first synced node by downloading the pre-compiled 1.4.1.4-jar instead of self-building from git checkout.
I followed these instructions: https://github.com/verydanny/iotaGoogleCloudInstance/wiki/Installing-IRI

**Comment by Pomyk** [comment]
I am debugging this issue. What I found so far is that the node receives a transaction, but somehow it doesn't get saved properly. Then it keeps requesting this transaction, but it still doesn't get saved. Maybe because it's in the received bytes/hashes cache?
That missing transaction stops the node from advancing solid subtangle milestone.

**Comment by romansemko** [comment]
I have two identical nodes. Same hardware, same (cloned) disk with installed IRI+Nelson, same static neighbors. One is always synced the other gets unsynced after few hours of restarting. I have noticed that the problematic one has lots of these logs:

`java[28773]: 12/30 19:08:43.570 [Solid Milestone Tracker] ERROR com.iota.iri.Milestone - Error during Solid Milestone updating`

Whereas the good one does not have any of these. So the problem seems to appear during updating of the solid milestones, which leads to sync lag. Can you guys check if you see any of these in your logs as well?

`journalctl -u iota.service | grep -i "Error during Solid Milestone updating" | less -N`

**Comment by nayr1991** [comment]
I don't see that in my logs. Here's the logs surrounding the last entry for Solid Milestone Tracker when it stopped syncing this week(last update at Dec 27 12:16:33):

```
Dec 27 12:14:40 #### java[28400]: 12/27 12:14:40.446 [Solid Milestone Tracker] INFO  com.iota.iri.Milestone - Latest SOLID SUBTANGLE milestone has changed from #311006 to #311011
Dec 27 12:14:43 #### java[28400]: 12/27 12:14:43.492 [Solid Milestone Tracker] INFO  com.iota.iri.Milestone - Latest SOLID SUBTANGLE milestone has changed from #311011 to #311012
Dec 27 12:14:46 #### java[28400]: 12/27 12:14:46.709 [pool-2-thread-2] INFO  com.iota.iri.network.Node - toProcess = 0 , toBroadcast = 0 , toRequest = 3 , toReply = 0 / totalTransactions = 6643965
Dec 27 12:14:48 #### java[28400]: 12/27 12:14:48.277 [Solid Milestone Tracker] INFO  com.iota.iri.Milestone - Latest SOLID SUBTANGLE milestone has changed from #311012 to #311013
Dec 27 12:14:56 #### java[28400]: 12/27 12:14:56.710 [pool-2-thread-2] INFO  com.iota.iri.network.Node - toProcess = 0 , toBroadcast = 0 , toRequest = 6 , toReply = 0 / totalTransactions = 6644000
Dec 27 12:15:06 #### java[28400]: 12/27 12:15:06.711 [pool-2-thread-2] INFO  com.iota.iri.network.Node - toProcess = 0 , toBroadcast = 0 , toRequest = 3 , toReply = 0 / totalTransactions = 6644068
Dec 27 12:15:09 #### java[28400]: 12/27 12:15:09.535 [Solid Milestone Tracker] INFO  com.iota.iri.Milestone - Latest SOLID SUBTANGLE milestone has changed from #311013 to #311015
Dec 27 12:15:16 #### java[28400]: 12/27 12:15:16.712 [pool-2-thread-2] INFO  com.iota.iri.network.Node - toProcess = 0 , toBroadcast = 0 , toRequest = 4 , toReply = 0 / totalTransactions = 6464841
Dec 27 12:15:18 #### java[28400]: 12/27 12:15:18.545 [Solid Milestone Tracker] INFO  com.iota.iri.Milestone - Latest SOLID SUBTANGLE milestone has changed from #311015 to #311017
Dec 27 12:15:23 #### java[28400]: 12/27 12:15:23.478 [Solid Milestone Tracker] INFO  com.iota.iri.Milestone - Latest SOLID SUBTANGLE milestone has changed from #311017 to #311018
Dec 27 12:15:26 #### java[28400]: 12/27 12:15:26.713 [pool-2-thread-2] INFO  com.iota.iri.network.Node - toProcess = 0 , toBroadcast = 0 , toRequest = 4 , toReply = 0 / totalTransactions = 6464889
Dec 27 12:15:29 #### java[28400]: 12/27 12:15:29.074 [Solid Milestone Tracker] INFO  com.iota.iri.Milestone - Latest SOLID SUBTANGLE milestone has changed from #311018 to #311020
Dec 27 12:15:33 #### java[28400]: 12/27 12:15:33.814 [Solid Milestone Tracker] INFO  com.iota.iri.Milestone - Latest SOLID SUBTANGLE milestone has changed from #311020 to #311023
Dec 27 12:15:36 #### java[28400]: 12/27 12:15:36.752 [XNIO-1 task-19] INFO  com.iota.iri.service.API - Removing neighbor: udp://188.194.143.121:14600
Dec 27 12:15:36 #### java[28400]: 12/27 12:15:36.755 [XNIO-1 task-18] INFO  com.iota.iri.service.API - Removing neighbor: udp://188.194.143.121:14600
Dec 27 12:15:36 #### java[28400]: 12/27 12:15:36.761 [XNIO-1 task-20] INFO  com.iota.iri.service.API - Adding neighbor: udp://52.48.68.100:14600
Dec 27 12:15:41 #### java[28400]: 12/27 12:15:41.714 [pool-2-thread-2] INFO  com.iota.iri.network.Node - toProcess = 0 , toBroadcast = 0 , toRequest = 6 , toReply = 0 / totalTransactions = 6464976
Dec 27 12:15:48 #### java[28400]: 12/27 12:15:48.744 [Solid Milestone Tracker] INFO  com.iota.iri.Milestone - Latest SOLID SUBTANGLE milestone has changed from #311023 to #311024
Dec 27 12:15:53 #### java[28400]: 12/27 12:15:53.322 [Solid Milestone Tracker] INFO  com.iota.iri.Milestone - Latest SOLID SUBTANGLE milestone has changed from #311024 to #311025
Dec 27 12:15:56 #### java[28400]: 12/27 12:15:56.715 [pool-2-thread-2] INFO  com.iota.iri.network.Node - toProcess = 0 , toBroadcast = 0 , toRequest = 4 , toReply = 0 / totalTransactions = 6465022
Dec 27 12:15:58 #### java[28400]: 12/27 12:15:58.522 [Solid Milestone Tracker] INFO  com.iota.iri.Milestone - Latest SOLID SUBTANGLE milestone has changed from #311025 to #311026
Dec 27 12:16:03 #### java[28400]: 12/27 12:16:03.446 [Solid Milestone Tracker] INFO  com.iota.iri.Milestone - Latest SOLID SUBTANGLE milestone has changed from #311026 to #311027
Dec 27 12:16:06 #### java[28400]: 12/27 12:16:06.716 [pool-2-thread-2] INFO  com.iota.iri.network.Node - toProcess = 0 , toBroadcast = 0 , toRequest = 6 , toReply = 0 / totalTransactions = 6465029
Dec 27 12:16:08 #### java[28400]: 12/27 12:16:08.687 [Solid Milestone Tracker] INFO  com.iota.iri.Milestone - Latest SOLID SUBTANGLE milestone has changed from #311027 to #311028
Dec 27 12:16:14 #### java[28400]: 12/27 12:16:14.242 [Solid Milestone Tracker] INFO  com.iota.iri.Milestone - Latest SOLID SUBTANGLE milestone has changed from #311028 to #311031
Dec 27 12:16:17 #### java[28400]: 12/27 12:16:17.118 [XNIO-1 task-30] INFO  com.iota.iri.service.API - Removing neighbor: udp://52.57.235.225:14600
Dec 27 12:16:19 #### java[28400]: 12/27 12:16:19.070 [Solid Milestone Tracker] INFO  com.iota.iri.Milestone - Latest SOLID SUBTANGLE milestone has changed from #311031 to #311033
Dec 27 12:16:21 #### java[28400]: 12/27 12:16:21.717 [pool-2-thread-2] INFO  com.iota.iri.network.Node - toProcess = 0 , toBroadcast = 0 , toRequest = 1 , toReply = 0 / totalTransactions = 6465153
Dec 27 12:16:25 #### java[28400]: 12/27 12:16:25.509 [XNIO-1 task-32] INFO  com.iota.iri.service.API - Adding neighbor: udp://45.77.232.81:14600
Dec 27 12:16:27 #### java[28400]: 12/27 12:16:27.202 [Latest Milestone Tracker] INFO  com.iota.iri.Milestone - Latest milestone has changed from #312563 to #312564
Dec 27 12:16:28 #### java[28400]: 12/27 12:16:28.906 [Solid Milestone Tracker] INFO  com.iota.iri.Milestone - Latest SOLID SUBTANGLE milestone has changed from #311033 to #311035
Dec 27 12:16:33 #### java[28400]: 12/27 12:16:33.905 [Solid Milestone Tracker] INFO  com.iota.iri.Milestone - Latest SOLID SUBTANGLE milestone has changed from #311035 to #311038
Dec 27 12:16:36 #### java[28400]: 12/27 12:16:36.717 [pool-2-thread-2] INFO  com.iota.iri.network.Node - toProcess = 0 , toBroadcast = 0 , toRequest = 6 , toReply = 0 / totalTransactions = 6465313
Dec 27 12:16:46 #### java[28400]: 12/27 12:16:46.718 [pool-2-thread-2] INFO  com.iota.iri.network.Node - toProcess = 0 , toBroadcast = 0 , toRequest = 8 , toReply = 0 / totalTransactions = 6465339
Dec 27 12:16:56 #### java[28400]: 12/27 12:16:56.719 [pool-2-thread-2] INFO  com.iota.iri.network.Node - toProcess = 0 , toBroadcast = 0 , toRequest = 8 , toReply = 0 / totalTransactions = 6465364
Dec 27 12:16:58 #### java[28400]: 12/27 12:16:58.368 [XNIO-1 task-40] INFO  com.iota.iri.service.API - Removing neighbor: udp://148.251.218.130:13370
Dec 27 12:17:02 #### java[28400]: 12/27 12:17:02.310 [XNIO-1 task-42] INFO  com.iota.iri.service.API - Removing neighbor: udp://45.77.232.81:14600
Dec 27 12:17:05 #### java[28400]: 12/27 12:17:05.812 [XNIO-1 task-43] INFO  com.iota.iri.service.API - Removing neighbor: udp://5.39.94.55:14600
Dec 27 12:17:07 #### java[28400]: 12/27 12:17:07.865 [XNIO-1 task-44] INFO  com.iota.iri.service.API - Removing neighbor: udp://37.120.168.112:14600
Dec 27 12:17:10 #### java[28400]: 12/27 12:17:10.539 [XNIO-1 task-47] INFO  com.iota.iri.service.API - Removing neighbor: udp://78.55.122.123:14777
Dec 27 12:17:11 #### java[28400]: 12/27 12:17:11.720 [pool-2-thread-2] INFO  com.iota.iri.network.Node - toProcess = 0 , toBroadcast = 0 , toRequest = 8 , toReply = 0 / totalTransactions = 6465381
Dec 27 12:17:16 #### java[28400]: 12/27 12:17:16.408 [XNIO-1 task-50] INFO  com.iota.iri.service.API - Removing neighbor: udp://185.207.106.172:14600
Dec 27 12:17:16 #### java[28400]: 12/27 12:17:16.604 [XNIO-1 task-49] INFO  com.iota.iri.service.API - Removing neighbor: udp://173.249.19.44:14600
Dec 27 12:17:20 #### java[28400]: 12/27 12:17:20.639 [XNIO-1 task-52] INFO  com.iota.iri.service.API - Removing neighbor: udp://111.231.86.86:14600
Dec 27 12:17:26 #### java[28400]: 12/27 12:17:26.029 [XNIO-1 task-53] INFO  com.iota.iri.service.API - Removing neighbor: udp://94.130.172.75:14600
Dec 27 12:17:26 #### java[28400]: 12/27 12:17:26.172 [XNIO-1 task-54] INFO  com.iota.iri.service.API - Removing neighbor: udp://40.70.59.44:14703
Dec 27 12:17:26 #### java[28400]: 12/27 12:17:26.721 [pool-2-thread-2] INFO  com.iota.iri.network.Node - toProcess = 0 , toBroadcast = 0 , toRequest = 7 , toReply = 0 / totalTransactions = 6465405
Dec 27 12:17:28 #### java[28400]: 12/27 12:17:28.048 [XNIO-1 task-56] INFO  com.iota.iri.service.API - Removing neighbor: udp://40.70.59.46:14600
Dec 27 12:17:28 #### java[28400]: 12/27 12:17:28.050 [XNIO-1 task-57] INFO  com.iota.iri.service.API - Removing neighbor: udp://40.70.59.46:14600
Dec 27 12:17:36 #### java[28400]: 12/27 12:17:36.524 [XNIO-1 task-58] INFO  com.iota.iri.service.API - Removing neighbor: udp://85.214.116.79:14600
Dec 27 12:17:36 #### java[28400]: 12/27 12:17:36.721 [pool-2-thread-2] INFO  com.iota.iri.network.Node - toProcess = 0 , toBroadcast = 0 , toRequest = 12 , toReply = 0 / totalTransactions = 6465437
Dec 27 12:17:37 #### java[28400]: 12/27 12:17:37.197 [XNIO-1 task-60] INFO  com.iota.iri.service.API - Removing neighbor: udp://173.212.221.15:14600
Dec 27 12:17:40 #### java[28400]: 12/27 12:17:40.832 [XNIO-1 task-63] INFO  com.iota.iri.service.API - Removing neighbor: udp://165.227.185.157:14600
Dec 27 12:17:46 #### java[28400]: 12/27 12:17:46.340 [XNIO-1 task-1] INFO  com.iota.iri.service.API - Removing neighbor: udp://178.254.35.151:14621
Dec 27 12:17:46 #### java[28400]: 12/27 12:17:46.365 [XNIO-1 task-2] INFO  com.iota.iri.service.API - Removing neighbor: udp://93.104.208.105:14600
Dec 27 12:17:46 #### java[28400]: 12/27 12:17:46.722 [pool-2-thread-2] INFO  com.iota.iri.network.Node - toProcess = 0 , toBroadcast = 0 , toRequest = 8 , toReply = 0 / totalTransactions = 6465472
```

**Comment by flqw** [comment]
I have a node running since about a week ago forked from the current dev branch and with some additional merged pull requests. The NullPointerExceptions do not occur in that version, but the solid milestone still gets stuck sometimes randomly. I can confirm the observation from @Pomyk that it seems to correlate with missing transactions: In my monitoring the numberOfTransactionsToRequest normally hovers around in the low single digit numbers and always goes back to zero. When the solid milestone gets stuck, the number is always at least 1 and slowly increasing.

Can somebody verify that the stuck milestone is always accompanied by the NullReferenceException in the current Release version? The Pull Request that fixed the Exception  https://github.com/iotaledger/iri/pull/431 might just have pushed the Problem to another location (probably the persistence?).

**Comment by malfidev** [comment]
Here is some debug info. Hope it helps to find the root cause:

My latest solid is stuck at
315334

While latest milestone currently is 
index: 315929
hash   : XDMTLHDGGYBRPVPWVSB9MYMVLAPJWRJFZGAWVNZUTS9DLAATLUILSBUPJGNSHHQDSFCTLWAXGQDDZ9999

In the debug console I see the following when it checks if the next milestone (315335) is solid

index : 315335
hash  : BYYMBJKJNNXMEENPFNBDETPWLKLSDRWAILGNAOSGDZSJDEYVWLXY9MKFUUBSFKLWDCISRGUHUVGWZ9999
Bytes : [-52, 75, 22, 64, -50, -121, 105, -65, -115, -92, -119, 37, 59, -115, 107, 92, 58, -80, 23, 27, 67, 14, 45, 60, -23, 57, 39, -49, -96, 107, -57, 81, 103, 87, -100, -71, 60, -44, 35, 3, -23, 60, -33, 64, 61, -31, 0, 0, 0]
HashCode: 669740659
Trtits: [-1, 1, 0, 1, -1, 0, 1, -1, 0, 1, 1, 1, -1, 1, 0, 1, 0, 1, -1, 1, 1, 1, 0, 1, -1, -1, -1, -1, -1, -1, 0, -1, 0, 1, 1, 1, -1, -1, 1, -1, -1, 1, -1, -1, -1, 1, -1, -1, 0, -1, 1, -1, -1, -1, -1, 1, 0, 1, 1, 0, -1, -1, 1, -1, 1, -1, 1, -1, -1, -1, -1, 0, 0, 1, 1, -1, 1, 1, 0, 1, 1, 1, 0, -1, 1, 1, 0, 0, 0, -1, -1, -1, 0, 1, 0, 0, 0, 0, 1, 0, 1, 1, 1, -1, 1, -1, -1, -1, 1, 0, 0, 0, -1, -1, 1, 0, -1, 1, -1, 1, 1, 1, 0, -1, 0, 0, 1, 0, -1, 1, 0, 1, 1, 1, 0, -1, -1, 1, 1, -1, 0, 1, 1, -1, -1, -1, 0, 0, 1, 1, 0, -1, 0, 1, -1, 0, 0, 0, 0, 1, 1, 1, -1, 1, 1, 0, -1, 1, 0, 1, -1, 0, 1, -1, -1, 1, 0, 1, 0, -1, 0, -1, 1, -1, 1, 1, 0, 1, 1, -1, -1, 0, 1, 1, 0, 0, 1, 0, 0, 0, 1, 1, 0, -1, 0, 0, -1, 1, -1, 1, 0, 1, -1, -1, 0, 1, 0, 1, -1, 1, 1, -1, 1, -1, 1, -1, -1, 0, -1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]

The issue is that 

public boolean checkSolidity(Hash hash, boolean milestone) (in Milestone.java)

always returns false for the solidity of 315335

During the debug session the latest milestone advanced 15 steps to 315944 while 315334 could not advance as the solidity for 315335 was still always false.

**Comment by alon-e** [comment]
>I am debugging this issue. What I found so far is that the node receives a transaction, but somehow it doesn't get saved properly. Then it keeps requesting this transaction, but it still doesn't get saved. Maybe because it's in the received bytes/hashes cache?
That missing transaction stops the node from advancing solid subtangle milestone.

@Pomyk that's a great find!
in this case, where the tx "somehow it doesn't get saved properly", the LRUcache would drop any further copy of the tx, as it "thinks" it was already saved (and b/c it's an LRU the tx will never leave the cache..) 

*this issue was spotted previously, but was, prob. falsely, associated w/ the new inflow rate limit.

I'll prepare a PR that bypasses the LRUcache altogether so others having this issue can evaluate if it resolves their situation (I haven't been able to reprod. this bug - after a lot of efforts to limit system specs).

meanwhile the underlying regression has yet to be found (why it doesn't get saved properly).

in any case having a FIFO cache (instead of LRU) may make more sense, allowing nodes to also "forget" such situations.

**Comment by Pomyk** [comment]
Made a PR with a workaround, maybe someone can test it.

**Comment by jdevelop** [comment]
@Pomyk will test that

**Comment by sreeraj-krishnan** [comment]
I will also try to test..

**Comment by jdevelop** [comment]
Doesn't work

```
{                                                                                                                                                                         [160/188]
  "appName": "IRI",                                                                                                                                            
  "appVersion": "1.4.1.4",                                                                                             
  "jreAvailableProcessors": 2,                                                                                                                                 
  "jreFreeMemory": 1090218608,                                                                                         
  "jreVersion": "1.8.0_152",                                                                                                                                   
  "jreMaxMemory": 3711959040,                                                                                          
  "jreTotalMemory": 1621622784,                                              
  "latestMilestone": "MKNDGPACVWGXURNPCBQPYHCOYLDDOHVCPML9SHJGEDMGNMQXFUYUP9CWZEHNCDNOADEYYSYTTJN9Z9999",              
  "latestMilestoneIndex": 316364,                                                                                                                              
  "latestSolidSubtangleMilestone": "MKEENIYOHSYOSISSLBFZNGHFIHQMLGDTIJLBHKOAD9MCQ9QCQXUYIKBCPHGMGBWRYIYFKUTIOCOTZ9999",
  "latestSolidSubtangleMilestoneIndex": 315906,                                                                                                                
  "neighbors": 8,                                                              
  "packetsQueueSize": 0,                                                                                                                                       
  "time": 1514769770243,                                                       
  "tips": 47,                                                                                                          
  "transactionsToRequest": 0,
  "duration": 1
}
```

and 2 hours later

```
{
  "appName": "IRI",
  "appVersion": "1.4.1.4",
  "jreAvailableProcessors": 2,
  "jreFreeMemory": 2223316584,
  "jreVersion": "1.8.0_152",
  "jreMaxMemory": 4046454784,
  "jreTotalMemory": 4046454784,
  "latestMilestone": "RUQ9GRHW9FRAZ9AKEDXDSVOSOWYQIXQZWXKHJIOOOL9YRCKIRSUPARPPAWAYHKOBHGP9IITISNIDZ9999",
  "latestMilestoneIndex": 316441,
  "latestSolidSubtangleMilestone": "MKEENIYOHSYOSISSLBFZNGHFIHQMLGDTIJLBHKOAD9MCQ9QCQXUYIKBCPHGMGBWRYIYFKUTIOCOTZ9999",
  "latestSolidSubtangleMilestoneIndex": 315906,
  "neighbors": 8,
  "packetsQueueSize": 0,
  "time": 1514778177416,
  "tips": 1404,
  "transactionsToRequest": 0,
  "duration": 0
}
```

**Comment by alon-e** [comment]
@jdevelop if your `TransactionToRequest` reaches 0, then you don't have the issue pomyk is dealing with.
Is it steady at 0/ most of the Time? 

The issue pomyk is dealing with is where your node falsely marks a transaction as stored (seen), then continues to request it, but does not store any new copy, b/c it's (falsely) marked stored.
In this case the `TransactionToRequest` would hover at least at 1 all the time.

**Comment by romansemko** [comment]
Testing @Pomyk solution as well...

**Comment by romansemko** [comment]
```
{
  "appName": "IRI",
  "appVersion": "1.4.1.4",
  "jreAvailableProcessors": 3,
  "jreFreeMemory": 909621576,
  "jreVersion": "1.8.0_151",
  "jreMaxMemory": 4772593664,
  "jreTotalMemory": 3465019392,
  "latestMilestone": "PLPWIIOFYFV9OBMLNITQRKBXOIMYAIKRUQAUKCRUUDQHPMX9VZUBNUOKICTFPNKBWNXFJXSZIIGLA9999",
  "latestMilestoneIndex": 316660,
  "latestSolidSubtangleMilestone": "LMMUITTJTCIBDUQJJMZBBOAFMPDJELG9XQNATSYDYTWXLGENDPATSQKKY9RTXZUFPEEPUCQTCTDEA9999",
  "latestSolidSubtangleMilestoneIndex": 316591,
  "neighbors": 32,
  "packetsQueueSize": 0,
  "time": 1514803616256,
  "tips": 450,
  "transactionsToRequest": 1,
  "duration": 0
}
```

**Comment by romansemko** [comment]
Seems to still be stuck:

```
{
  "appName": "IRI",
  "appVersion": "1.4.1.4",
  "jreAvailableProcessors": 3,
  "jreFreeMemory": 1269376200,
  "jreVersion": "1.8.0_151",
  "jreMaxMemory": 4772593664,
  "jreTotalMemory": 3990880256,
  "latestMilestone": "LKJIZQO9HUGLPVWBMTWEOJFZOVOZACEPPDZWCZHGZTHSLWBHXHAWYDLPYAZHVBEVQDOSHNC9YRNO99999",
  "latestMilestoneIndex": 316678,
  "latestSolidSubtangleMilestone": "LMMUITTJTCIBDUQJJMZBBOAFMPDJELG9XQNATSYDYTWXLGENDPATSQKKY9RTXZUFPEEPUCQTCTDEA9999",
  "latestSolidSubtangleMilestoneIndex": 316591,
  "neighbors": 27,
  "packetsQueueSize": 0,
  "time": 1514805807458,
  "tips": 639,
  "transactionsToRequest": 1,
  "duration": 1
}
```

**Comment by brunoamancio** [comment]
@romansemko did you try cleaning mainnetdb and mainnet.log folders and redownload the db from http://iota.partners ? I've done it and it took a bit, but it synced :D

**Comment by romansemko** [comment]
@brunoamancio , you mean using the proposed fix after re-downloading db? I can try that..

**Comment by Pomyk** [comment]
I am running this on 3 nodes that were affected. One is running for 14 hours and two are running for 4 hours. So far all are synced.

**Comment by romansemko** [comment]
Okay, just got fresh DB install from iota.partners. Running again. *fingers crossed*

**Comment by brunoamancio** [comment]
@romansemko  yes, using the solution from @Pomyk with an almost synced DB from iota.partners.

**Comment by romansemko** [comment]
For the first time I see this node in sync in a very long time! Good job!

**Comment by jdevelop** [comment]
@alon-e right, I've added some debug statements to the master branch build and realized that in my case the latest milestone index stops advancing after catching fire at

```
INFO  com.iota.iri.Snapshot - Transaction resolves to incorrect ledger balance: -603112979334684
```

is there any hints on what might be wrong there? I'm digging through the code, and it's not very easy, so some help would be highly appreciated.

**Comment by alon-e** [comment]
@jdevelop it seems you have another issue than what is discussed in this issue (or at least the last N comments) - essentially your local database may have become corrupted, as that message (in the context of milestones) means that the accepting the next milestone would lead to an inconsistent ledger - hence the `latestSolidSubtangleMilestone` isn't progressing (`ledgerValidator.updateSnapshot(latestSnapshot, milestoneViewModel)` fails).

please try to restart your node w/ the `--revalidate` flag - if this doesn't work, you can try to re-sync using @brunoamancio's advice.

if this persists. please open a new, dedicated, issue - with more of info from the error log.

**Comment by romansemko** [comment]
I am still in sync. By far, the longest period on that node.

**Comment by thepill** [comment]
it also worked for me to recover from `[Solid Milestone Tracker] INFO  com.iota.iri.Snapshot - Transaction resolves to incorrect ledger balance: -100000000` using the db snapshot from http://iota.partners/index.html.

Back in sync after about 20 minutes 👍

**Comment by brunoamancio** [comment]
I'm still synced as well. First time in a long time for me too. 

The technical details of the solution should be checked merged. Also, a new IRI build should be made available asap, as there a great deal of nodes aren't synced due to that bug.

**Comment by romansemko** [comment]
I am still synced after 14 hours! Whoaaaa! 💃

**Comment by stockermans** [comment]
I think this may have fixed it, I used the modified file from Pomyk and compiled iri and have been running with it perhaps around 7 hours and synced.

**Comment by OMSKremer** [comment]
same here, 13 hours fully synched, longest time since node is alive. I think patch/workaround is confirmed. Thx @Pomyk !

**Comment by romansemko** [comment]
Almost 24 hours now. Still running smoothly. All nodes on the latest milestone.
I think this fix is confirmed, as well.

**Comment by alon-e** [comment]
good to hear!
thanks for the feedback, 
we have also been running some nodes with the patch that are keeping sync well.

I've been in contact w/ @Pomyk,
reviewed the workaround & will comment on the PR itself.

**Comment by nimearo** [comment]
On my side the changes from pull request #455 also look good so far. I currently use HEAD of  dev branch with changes from pull request #455 and the solid milestone still keeps updating after more then 8 hours. Never accomplished that before in the last month. First of all I thought the PR would not solve my problem. But after restarting iri with --revalidate solid milestone keeps moving up. Will update here when I reach latest milestone again.

1. Use HEAD of dev branch
2. Pull changes from #455
3. Start iri with --revalidate option.

**Comment by manuelbieh** [comment]
Have the same issue here. Node was running very well for a week or two but since a few days it's always out of sync, even directly after resynching it. That's really frustrating. My machine has 12 GB RAM and a Quadcore Xeon CPU. And I'm running Nelson.

**Comment by sreeraj-krishnan** [comment]
#459 is not working for me.

**Comment by brunoamancio** [comment]
if you wanne be synced for now, I suggest https://github.com/Pomyk/iri/commit/041fbbecb5c6c64fec036beab65cfafdce9f6ea7

**Comment by michi6383** [comment]
@Pomyk, can confirm the workaround is doing it's job. running for 7,5h now. Thank you!

**Comment by sreeraj-krishnan** [comment]
one of my nodes is sync for more than 24hrs with pomyk fix

**Comment by malfidev** [comment]
Yep, can confirm that Pomyk fix does work fine for more than 2 days. Also the other values (tips/transactions to request) look healthy. Thanks for the fix.

**Comment by chrishamm** [comment]
I can confirm @Pomyk's fix works - huge thanks for that! This is the first time my node has been in sync for more than 24 hours.

**Comment by GhostTyper** [comment]
I installed @Pomyk 's fix on node.lukaseder.de. This node is a heavy loaded public wallet node (avg of 350 concurrent clients). This fix "crashed" the node two times, so I rolled back.

The "crash" made the API access impossible but didn't result in a segfault or java panic.

After rollback the node workes just fine again.

**Comment by vctender** [comment]
@GhostTyper Maybe your problem is caused by other changes from current dev branch?! There are only a few changes made by Pomyk. Have you checked system logs for any errors?

**Comment by GhostTyper** [comment]
Well, I don't say, that this error is due to @Pomyk s changes. But i can say it ocurred with his version.

**Comment by jacklc** [comment]
I still have this problem after IRI update.

**Comment by lekanovic** [comment]
I had the same problem, I am trying to delete the DB and download the latest one to see if that solves the problem

**Comment by hapklaar** [comment]
Exception in thread "Solid Milestone Tracker" still present in 1.4.1.7 on linux
